homeassistant/components/homeconnect/switch.py

```diff
@@ -0,0 +1,189 @@
+"""Provides a switch for Home Connect.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/integrations/switch.homeconnect/
+"""
+import logging
+import re
+
+from homeconnect.api import HomeConnectError
+
+from homeassistant.components.switch import SwitchDevice
+
+from .api import HomeConnectEntity
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ["homeconnect"]
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Home Connect switch."""
+
+    def get_entities():
+        """Get a list of entities."""
+        entities = []
+        hc_api = hass.data[DOMAIN][config_entry.entry_id]
+        for device_dict in hc_api.devices:
+            entity_dicts = device_dict.get("entities", {}).get("switch", [])
+            entity_list = [HomeConnectProgramSwitch(**d) for d in entity_dicts]
+            entity_list += [HomeConnectPowerSwitch(device_dict["device"])]
+            device = device_dict["device"]
+            device.entities += entity_list
+            entities += entity_list
+        return entities
+
+    async_add_entities(await hass.async_add_executor_job(get_entities), True)
+
+
+class HomeConnectProgramSwitch(HomeConnectEntity, SwitchDevice):
+    """Switch class for Home Connect."""
+
+    def __init__(self, device, program_name):
+        """Initialize the entity."""
+        name = " ".join([device.appliance.name, "Program", program_name.split(".")[-1]])
+        super().__init__(device, name)
+        self.program_name = program_name
+        self._state = None
+        self._remote_allowed = None
+
+    @property
+    def is_on(self):
+        """Return true if the switch is on."""
+        return bool(self._state)
+
+    @property
+    def available(self):
+        """Return true if the entity is available."""
+        return True
+
+    def turn_on(self, **kwargs):
+        """Start the program."""
+        _LOGGER.debug("tried to turn on program %s", self.program_name)
+        try:
+            self.device.appliance.start_program(self.program_name)
+        except HomeConnectError as err:
+            _LOGGER.error("Error while trying to start program: %s", err)
+        self.async_entity_update()
```

Comment:
We can't call an async function from a sync context. Then we need to use `hass.add_job` to schedule the async work from the sync context.

Or change `turn_on` to the async version and schedule the sync work on the executor thread pool with `hass.async_add_executor_job`.

It's generally better to have the outer context be async and schedule inner things on the executor as needed, while trying to do that in as large chunks as possible to minimize context switching.

Reply:
OK thanks for the hint, can you help me out a little more since I'm not good at async?
- If I start the program with `async_add_executor_job` and `await` that, is the `async_entity_update` is guaranteed to be called *after* the job has been executed? (This is crucial as the state of the switch needs to be updated after making the API call.)
- Does it make sense to have the `await hass.async_add_executor_job` inside the `try`...`except` or how do I otherwise log the traceback of the sync API call?

Reply:
1. Yes.
2. Yes, the try... except should wrap the executor job, but nothing more.